### ISSUE TITLE
Possible PrefixQuant issues on Smaller LLMs?

### ISSUE THREAD

**Issue by sasha-hailo** [ChenMnZ/PrefixQuant#23]
Dear @ChenMnZ,
Thank you for the paper and sharing the code.
I'm very interested about the idea. Especially encouraging is the prospect of reaching great accuracy using static per-channel / tensor quantization.

I evaluated PrefixQuant with a few HF models, and was able to reproduce the good results of your paper with `Llama-2-7B`, `Llama-3-8B`, `Mistral-7B` (and a couple other models).
_[I'm writing this to indicate that my settings are probably OK, and are not the cause for the issues below]_

However, with smaller models (of sizes 0.5B - 3B), the results I'm getting are catastrophic (multi-digit perplexity numbers on WikiText2).
The models I tested that resulted in PPL disasters are:
- `Qwen2-0.5B`
- `Llama-3.2-1B`
- `Qwen2-1.5B-Instruct`
- `Llama-3.2-3B-Instruct`

A couple remarks:
- `Llama-3.2-xx` models are not supported by the older `transformers` library version required by your repo (4.40.1). 
I encountered minor compatibility issues when trying to run your repo with a more recent `transformers` version, but solved them by explicitly configuring `use_cache=True`. But even after solving them, the models would not properly quantize.
- For `Qwen2-0.5B` model, there was an issue with creating the online Hadamard matrix for `down_proj` input (function `get_hadK()` does not support the intermediate feature size). I overcame it by disabling the configuration option `down_online_had`, but it didn't help reach reasonable quantization accuracy.
- As a sanity check, I tried to quantize one of the problematic models listed above with additional fine-tuning - this did not help, either.

Do you have any inputs that can help SLM models quantize well?

Thanks in advance!

**Comment by sasha-hailo** [comment]
Nailed it!
After comparing with another quantization repo implementing rotation (LLMC),
I found the code that makes it work.
Please see lines 23-33 here:
https://github.com/ModelTC/llmc/blob/9943e792f5f49341351dbd0c242ce8e0ba50b9a9/llmc/compression/quantization/quarot.py#L26

**Comment by ChenMnZ** [comment]
@sasha-hailo Thanks for reporting this issue and offer a solution!

**Comment by ChenMnZ** [comment]
@sasha-hailo 
If you suffer performance degeneration after fine-tuning, you can try to add `--loss_type skip_mse` for more stable training.

Additionally, I also observed that prefixed token may significantly increase the perplexity of some small model even without quantization. In this case, you should test on more other tasks to analusis weather prefixed tokens are useable.

**Comment by a19284** [comment]
> Nailed it! After comparing with another quantization repo implementing rotation (LLMC), I found the code that makes it work. Please see lines 23-33 here: https://github.com/ModelTC/llmc/blob/9943e792f5f49341351dbd0c242ce8e0ba50b9a9/llmc/compression/quantization/quarot.py#L26

@sasha-hailo can you share the full patch please? I'd like to try prefixquant on Llama-3.2-3B-Instruct. thx

**Comment by sasha-hailo** [comment]
> > Nailed it! After comparing with another quantization repo implementing rotation (LLMC), I found the code that makes it work. Please see lines 23-33 here: https://github.com/ModelTC/llmc/blob/9943e792f5f49341351dbd0c242ce8e0ba50b9a9/llmc/compression/quantization/quarot.py#L26
> 
> [@sasha-hailo](https://github.com/sasha-hailo) can you share the full patch please? I'd like to try prefixquant on Llama-3.2-3B-Instruct. thx

@a19284 
In file `utils/rotation_utils.py`, function `fuse_layer_norms()`, I added the following patch in line 54:
```
    embeddings = model_utils.get_embeddings(**kwargs)[0]
    lm_head = model_utils.get_lm_head(**kwargs)
    if torch.equal(embeddings.weight, lm_head.weight):
        print("Tie weight! Copy embed_layer for head_layer!")
        model.config.tie_word_embeddings = False
        model._tie_or_clone_weights = lambda *args, **kwargs: None
        del lm_head.weight
        w = embeddings.weight.clone()
        lm_head.weight = torch.nn.Parameter(w)
```

Hope it helps!

**Comment by haooxia** [comment]
> > > Nailed it! After comparing with another quantization repo implementing rotation (LLMC), I found the code that makes it work. Please see lines 23-33 here: https://github.com/ModelTC/llmc/blob/9943e792f5f49341351dbd0c242ce8e0ba50b9a9/llmc/compression/quantization/quarot.py#L26
> > 
> > 
> > [@sasha-hailo](https://github.com/sasha-hailo) can you share the full patch please? I'd like to try prefixquant on Llama-3.2-3B-Instruct. thx
> 
> [@a19284](https://github.com/a19284) In file `utils/rotation_utils.py`, function `fuse_layer_norms()`, I added the following patch in line 54:
> 
> ```
>     embeddings = model_utils.get_embeddings(**kwargs)[0]
>     lm_head = model_utils.get_lm_head(**kwargs)
>     if torch.equal(embeddings.weight, lm_head.weight):
>         print("Tie weight! Copy embed_layer for head_layer!")
>         model.config.tie_word_embeddings = False
>         model._tie_or_clone_weights = lambda *args, **kwargs: None
>         del lm_head.weight
>         w = embeddings.weight.clone()
>         lm_head.weight = torch.nn.Parameter(w)
> ```
> 
> Hope it helps!

Hi,
I encountered the same issue on the LLaMA and Qwen small models.
With your patch, I can obtain a reasonable PPL on Llama-3.2-3B-Instruct, but the Qwen series (such as Qwen2-1.5B-Instruct) still gives very large PPL.
(In addition, I was also unable to reproduce the results in the paper on Qwen2-7B, although Llama3-8B works fine.)

I'm wondering if there might be some parameter settings or code adjustments specific to Qwen models that differ from LLaMA.
Below are the parameters I’m currently using.

Thank you very much for your time and help!


```
python main.py \
--model_path path_to/Qwen2-1.5B-Instruct/  \
--model_name qwen-2-1.5b \
--output_dir ./log/tmp \
--wbits 4 \
--input_bits 4 \
--input_mode static \
--v_bits 4 \
--k_bits 4 \
--kv_group_size 128 \
--kv_mode static \
--mse_init \
--pre_rotate \
--down_online_had \
--qk_online_had \
--set_prefixed_tokens \
--eval_ppl \
--eval_tasks  piqa,arc_easy,arc_challenge,hellaswag,winogrande \
--save_quant_dir ./pre_quantized_models/tmp
```

**Comment by sasha-hailo** [comment]
> > > > Nailed it! After comparing with another quantization repo implementing rotation (LLMC), I found the code that makes it work. Please see lines 23-33 here: https://github.com/ModelTC/llmc/blob/9943e792f5f49341351dbd0c242ce8e0ba50b9a9/llmc/compression/quantization/quarot.py#L26
> > > 
> > > 
> > > [@sasha-hailo](https://github.com/sasha-hailo) can you share the full patch please? I'd like to try prefixquant on Llama-3.2-3B-Instruct. thx
> > 
> > 
> > [@a19284](https://github.com/a19284) In file `utils/rotation_utils.py`, function `fuse_layer_norms()`, I added the following patch in line 54:
> > ```
> >     embeddings = model_utils.get_embeddings(**kwargs)[0]
> >     lm_head = model_utils.get_lm_head(**kwargs)
> >     if torch.equal(embeddings.weight, lm_head.weight):
> >         print("Tie weight! Copy embed_layer for head_layer!")
> >         model.config.tie_word_embeddings = False
> >         model._tie_or_clone_weights = lambda *args, **kwargs: None
> >         del lm_head.weight
> >         w = embeddings.weight.clone()
> >         lm_head.weight = torch.nn.Parameter(w)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hope it helps!
> 
> Hi, I encountered the same issue on the LLaMA and Qwen small models. With your patch, I can obtain a reasonable PPL on Llama-3.2-3B-Instruct, but the Qwen series (such as Qwen2-1.5B-Instruct) still gives very large PPL. (In addition, I was also unable to reproduce the results in the paper on Qwen2-7B, although Llama3-8B works fine.)
> 
> I'm wondering if there might be some parameter settings or code adjustments specific to Qwen models that differ from LLaMA. Below are the parameters I’m currently using.
> 
> Thank you very much for your time and help!
> 
> ```
> python main.py \
> --model_path path_to/Qwen2-1.5B-Instruct/  \
> --model_name qwen-2-1.5b \
> --output_dir ./log/tmp \
> --wbits 4 \
> --input_bits 4 \
> --input_mode static \
> --v_bits 4 \
> --k_bits 4 \
> --kv_group_size 128 \
> --kv_mode static \
> --mse_init \
> --pre_rotate \
> --down_online_had \
> --qk_online_had \
> --set_prefixed_tokens \
> --eval_ppl \
> --eval_tasks  piqa,arc_easy,arc_challenge,hellaswag,winogrande \
> --save_quant_dir ./pre_quantized_models/tmp
> ```

I'm not sure (and I'm not the code / paper author),
My guess would be that weight quantization is not the root cause of the high degradations.
In my experiments, Qwen2 models showed reasonable PPL in weight-only quantization scenarios.
Perhaps the degradation is due to `input_bits` or `k_bits` - I'd try and profile separately.

**Comment by haooxia** [comment]
> > > > > Nailed it! After comparing with another quantization repo implementing rotation (LLMC), I found the code that makes it work. Please see lines 23-33 here: https://github.com/ModelTC/llmc/blob/9943e792f5f49341351dbd0c242ce8e0ba50b9a9/llmc/compression/quantization/quarot.py#L26
> > > > 
> > > > 
> > > > [@sasha-hailo](https://github.com/sasha-hailo) can you share the full patch please? I'd like to try prefixquant on Llama-3.2-3B-Instruct. thx
> > > 
> > > 
> > > [@a19284](https://github.com/a19284) In file `utils/rotation_utils.py`, function `fuse_layer_norms()`, I added the following patch in line 54:
> > > ```
> > >     embeddings = model_utils.get_embeddings(**kwargs)[0]
> > >     lm_head = model_utils.get_lm_head(**kwargs)
> > >     if torch.equal(embeddings.weight, lm_head.weight):
> > >         print("Tie weight! Copy embed_layer for head_layer!")
> > >         model.config.tie_word_embeddings = False
> > >         model._tie_or_clone_weights = lambda *args, **kwargs: None
> > >         del lm_head.weight
> > >         w = embeddings.weight.clone()
> > >         lm_head.weight = torch.nn.Parameter(w)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Hope it helps!
> > 
> > 
> > Hi, I encountered the same issue on the LLaMA and Qwen small models. With your patch, I can obtain a reasonable PPL on Llama-3.2-3B-Instruct, but the Qwen series (such as Qwen2-1.5B-Instruct) still gives very large PPL. (In addition, I was also unable to reproduce the results in the paper on Qwen2-7B, although Llama3-8B works fine.)
> > I'm wondering if there might be some parameter settings or code adjustments specific to Qwen models that differ from LLaMA. Below are the parameters I’m currently using.
> > Thank you very much for your time and help!
> > ```
> > python main.py \
> > --model_path path_to/Qwen2-1.5B-Instruct/  \
> > --model_name qwen-2-1.5b \
> > --output_dir ./log/tmp \
> > --wbits 4 \
> > --input_bits 4 \
> > --input_mode static \
> > --v_bits 4 \
> > --k_bits 4 \
> > --kv_group_size 128 \
> > --kv_mode static \
> > --mse_init \
> > --pre_rotate \
> > --down_online_had \
> > --qk_online_had \
> > --set_prefixed_tokens \
> > --eval_ppl \
> > --eval_tasks  piqa,arc_easy,arc_challenge,hellaswag,winogrande \
> > --save_quant_dir ./pre_quantized_models/tmp
> > ```
> 
> I'm not sure (and I'm not the code / paper author), My guess would be that weight quantization is not the root cause of the high degradations. In my experiments, Qwen2 models showed reasonable PPL in weight-only quantization scenarios. Perhaps the degradation is due to `input_bits` or `k_bits` - I'd try and profile separately.

Yes, thank you very much for your reply.
I conducted separate experiments (based on Qwen-2-7B) and observed that the issue is indeed related to `k_bits`. When `k_bits` is set to 16, I can obtain reasonable PPL results (PPL=8.07).

I'll check with the authors in more detail on how to reproduce the W4A8KV4 results. Thanks again!
